lib/pure/strtabs.nim

```diff
@@ -272,6 +272,32 @@ proc `%`*(f: string, t: StringTableRef, flags: set[FormatFlag] = {}): string {.
       add(result, f[i])
       inc(i)
 
+proc del*(t: StringTableRef, key: string) =
+  ## Removes `key` from `t`.
+  # Impl adapted from `tableimpl.delImplIdx`
+  var i = rawGet(t, key)
+  let msk = high(t.data)
+  if i >= 0:
+    dec(t.counter)
+    block outer:
+      while true:
+        var j = i
+        var r = j
+        t.data[i].hasValue = false
+        t.data[i].key = ""
+        t.data[i].val = ""
+        while true:
+          i = ((5 * i) + 1) and msk    # increment mod table size
+          if not t.data[i].hasValue:   # end of collision cluster; So all done
+            break outer
+          r = t.myhash(t.data[i].val) and msk    # "home" location of key@i
+          if not ((i >= r and r > j) or (r > j and j > i) or (j > i and i >= r)):
+            break
+        when defined(js):
+          t.data[j] = t.data[i]
+        else:
+          shallowCopy(t.data[j], t.data[i]) # data[j] will be marked EMPTY next loop
```

Comment:
Offtopic, but wouldn't it make more sense to put the `when defined(js)` block in the shallowCopy proc?

Reply:
Seems like `shallowCopy` already works with js, so this check might just be legacy code (the code is taken from tableimpl.nim).